File: src/RNetPi.Core/RNet/SetPowerPacket.cs
using System;

namespace RNetPi.Core.RNet;

/// <summary>
/// Packet for setting zone power
/// </summary>
public class SetPowerPacket : EventPacket
{
    public SetPowerPacket(byte controllerID, byte zoneID, bool power)
    {
        TargetPath = new byte[] { 0x02, 0x00 };
        TargetControllerID = controllerID;
        EventID = power ? (ushort)0xDD : (ushort)0xDC;
        EventTimestamp = 0;
        EventData = zoneID;
        EventPriority = 1;
    }

    public byte GetControllerID()
    {
        return TargetControllerID;
    }

    public byte GetZoneID()
    {
        return (byte)EventData;
    }

    public bool GetPower()
    {
        return EventID == 0xDD;
    }
}

File: src_js/rnet/packets/SetPowerPacket.js
const EventPacket = require("./EventPacket");

class SetPowerPacket extends EventPacket {
    constructor(controllerID, zoneID, power) {
        super();

        this.targetPath = [0x02, 0x00];
        this.targetControllerID = controllerID;
        this.eventID = 0xDC;
        this.eventData = zoneID;
        this.eventTimestamp = power === true ? 1 : 0;
        this.eventPriority = 1;
    }

    getControllerID() {
        return this.targetControllerID;
    }

    getZoneID() {
        return this.eventData;
    }

    getPower() {
        return this.eventTimestamp == 1;
    }
}

SetPowerPacket.fromPacket = function(eventpacket) {
    if (eventPacket instanceof EventPacket) {
        const powerPacket = new SetPowerPacket();
        eventPacket.copyToPacket(powerPacket);
        return powerPacket;
    }
    else {
        throw new TypeError("Cannot create SetPowerPacket from anything other than an EventPacket");
    }
}

module.exports = SetPowerPacket;
